src/inference/involution_dsl.jl

```diff
@@ -0,0 +1,440 @@
+import ForwardDiff
+import MacroTools
+import LinearAlgebra
+
+# TODO allow providing a pair of address namespaces, one for the model and one
+# for the proposal, when calling another involution function with call
+
+# TODO proposal_retval is a way that continuous data can leak unaccounted for..
+# for now, the requirement is that the return value of f_disc cannot depend on
+# continuous addresses in the model or proposal although in the future, we
+# could do AD through its return value as well using choice_gradients() and add
+# these to the Jacobian...?
+
+# TODO add more checks that we don't visit the same address twice (these can be
+# only enabled in 'check' mode)
+
+struct InvolutionDSLProgram
+    fn!::Function
+end
+
+struct FirstPassState
+    trace
+    u::ChoiceMap
+    constraints::ChoiceMap
+    u_back::ChoiceMap
+    t_cont_reads::Dict
+    u_cont_reads::Dict
+    t_cont_writes::Dict
+    u_cont_writes::Dict
+    t_copy_reads::DynamicSelection
+    u_copy_reads::DynamicSelection
+    marked_as_retained::Set
+end
+
+function FirstPassState(trace, u::ChoiceMap)
+    FirstPassState(
+        trace, u, choicemap(), choicemap(),
+        Dict(), Dict(), Dict(), Dict(),
+        DynamicSelection(), DynamicSelection(),
+        Set())
+end
+
+struct JacobianPassState{T}
+    trace
+    u::ChoiceMap
+    input_arr::AbstractArray{T}
+    output_arr::Array{T}
+    t_key_to_index::Dict
+    u_key_to_index::Dict
+    cont_constraints_key_to_index::Dict
+    cont_u_back_key_to_index::Dict
+end
+
+const inv_state = gensym("inv_state")
+
+"""
+    @involution function f(...)
+        ..
+    end
+
+Write a program in the [Involution DSL](@ref).
+"""
+macro involution(ex)
+    MacroTools.@capture(ex, function f_(args__) body_ end) || error("expected syntax: function f(..) .. end")
+
+    fn! = gensym("$(esc(f))_fn!")
+
+    quote
+
+    # mutates the state
+    function $fn!($(esc(inv_state))::Union{FirstPassState,JacobianPassState}, $(map(esc, args)...))
+        $(esc(body))
+        nothing
+    end
+
+    $(esc(f)) = InvolutionDSLProgram($fn!)
```

Comment:
Would it be easy to tweak this so that `f` picks up any docstring provided via
```julia
"""My docstring"""
@involution my_involution ...
```
?  (Asking because I think you did similarly before with the `@gen` macro)

Reply:
Done